salt/orch/removal.sls

```diff
@@ -0,0 +1,146 @@
+{%- set target = salt['pillar.get']('target') %}
+
+{%- set masters = salt.saltutil.runner('mine.get', tgt='G@roles:kube-master', fun='network.interfaces', tgt_type='compound').keys() %}
+{%- set etcd_members = salt.saltutil.runner('mine.get', tgt='G@roles:etcd', fun='network.interfaces', tgt_type='compound').keys() %}
+
+sync_pillar:
+  salt.runner:
+    - name: saltutil.sync_pillar
+
+update_pillar:
+  salt.function:
+    - tgt: '*'
+    - name: saltutil.refresh_pillar
+    - require:
+      - salt: sync_pillar
+
+update_grains:
+  salt.function:
+    - tgt: '*'
+    - name: saltutil.refresh_grains
+
+update_mine:
+  salt.function:
+    - tgt: '*'
+    - name: mine.update
+    - require:
+       - salt: update_pillar
+       - salt: update_grains
+
+update_modules:
+  salt.function:
+    - name: saltutil.sync_modules
+    - tgt: '*'
+    - kwarg:
+        refresh: True
+    - require:
+      - salt: update_mine
+
+{% if target in etcd_members %}
+  # if the node is a member of the etcd cluster:
+  # we must choose another node and promote it to be an etcd master
+  {% set non_etcd_members = salt.saltutil.runner('mine.get', tgt='not G@roles:etcd', fun='network.interfaces', tgt_type='compound').keys() %}
+  {% if non_etcd_members|length > 0 %}
+    {% set new_etcd_member = non_etcd_members[0] %}
+
+    # promote some other node to be a full member of the etcd cluster
+    # WARNING: we should do a consistency check in Velum...
+    #          if no proxy is available for promotion and:
+    #          * num_members >= 3  ->  we should show a "are you sure??" message.
+    #          * num_members == 1  ->  we should just not allow it
+    #
+add_new_etcd_member:
+  salt.state:
+    - tgt: '{{ new_etcd_member }}'
+    - sls:
+      - etcd.remove-pre-stop-services
+    - require:
+      - salt: update_modules
+  {% endif %}
+
+  # remove the etcd master running in {{ target }}
+etc_remove_master:
+  salt.state:
+    - tgt: '{{ target }}'
+    - sls:
+      - etcd.remove-pre-stop-services
+    - require:
+      {%- if non_etcd_members|length > 0 %}
+      - salt: add_new_etcd_member
+      {%- else %}
+      - salt: update_modules
+      {%- endif %}
+
+{% endif %}
+
+# TODO: if {{ target }} is a k8s master, we should promote a worker... but this is
+#       too hard to do right now
+
+update_grains_again:
+  salt.function:
+    - tgt: '*'
+    - name: saltutil.refresh_grains
+    {%- if target in etcd_members %}
+    - require:
+      - salt: etc_remove_master
+    {%- endif %}
+
+stop_services:
+  salt.state:
+    - tgt: {{ target }}
+    - sls:
+      - container-feeder.stop
+      {%- if target in masters %}
+      - kube-apiserver.stop
+      - kube-controller-manager.stop
+      - kube-scheduler.stop
+      {%- else %}
+      - kubelet.stop
+      - kube-proxy.stop
+      {%- endif %}
+      - docker.stop
+      - etcd.stop
+    - require:
+      - salt: update_grains_again
+
+# revoke certificates
+# TODO
+
+# revoke Salt keys
```

Comment:
Hmmm I think that revoking salt keys should goes after reboot.

Reply:
IMO we shouldn't reboot - we should shutdown.. So the revokes probably belong here.

Reply:
Yeah, we should probably shutdown...